src/lava/magma/core/model/py/connection.py

```diff
@@ -631,36 +635,80 @@ def _init_randoms(self) -> None:
 
         self._conn_var_random = ConnVarRandom()
 
-    def _record_pre_spike_times(self, s_in: np.ndarray) -> None:
-        """Record within-epoch spiking times of pre- and post-synaptic neurons.
+    def _process_pre_spikes(self, s_in: np.ndarray) -> None:
+        """Process pre-synaptic spikes.
 
-        If more a single pre- or post-synaptic neuron spikes more than once,
-        the corresponding trace is updated by its trace impulse value.
+        Four different modes of operation, based on GradedSpikeCfg value:
+        (0) GradedSpikeCfg.DEFAULT if a single pre-synaptic neuron spikes more
+        than once, pre-traces are updated by their regular impulses.
+        (1) GradedSpikeCfg.OVERWRITE overwrites the value of the pre-synaptic
+        trace x1 by payload/2, upon spiking.
+        (2) GradedSpikeCfg.ADD_SATURATION adds payload/2 to the pre-synaptic
+        trace x1, upon spiking, saturates x1 to 127.
+        (3) GradedSpikeCfg.ADD_NO_SATURATION adds payload/2 to the pre-synaptic
+        trace x1, upon spiking, keeps only overflow from 127 in x1,
+        adds regular impulse to x2 on overflow.
+
+        Within-epoch spike times are recorded.
+        With GradedSpikeCfg.ADD_NO_SATURATION, only spike times of spikes
+        triggering x1 overflow and x2 impulse addition are recorded.
```

Comment:
This docstring is copied to several places within this file. Can that be defined once in a central location and just referenced elsewhere?

Reply:
It is actually similar but not exactly the same within this file.
There is a subtle difference between the FloatingPoint and the BitApprox versions when it comes to this docstring.

An option would be to define this docstring in the PyLearningConnection class (which is the parent class of both the FloatingPoint and BitApprox versions) might be an idea. However, the method in question (`_process_pre_spikes`) is abstract in PyLearningConnection and is only implemented in children classes, which makes me a bit reluctant to put the docstring there.